### ISSUE TITLE
`MongoError: Cannot create field '-1' in subelement` on `save()`

### ISSUE THREAD

**Issue by tevin** [Automattic/mongoose#7970]
**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**
When updating a document using `save()`, mongoose throws the error below: 
`Cannot create field '-1' in element {subdocument: ...}`

It seems like mongoose attempts to create a field in one of my subdocuments with the name as `-1`

I am not making any changes to the subdocument. 

**If the current behavior is a bug, please provide the steps to reproduce.**

 
Gist with relevant code snippets is [here](https://gist.github.com/tevin/c20e7c01d2028b794bda96d5fdeb57f6)

**What is the expected behavior?**
The save should run successfully without attempting to create a new field in the subdocuments.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node: 10.11.0
Mongoose: 5.6.0
MongoDB: 4.0.2

**Comment by tevin** [comment]
There's also a [similar issue on Stack Overflow](https://stackoverflow.com/questions/54437833/why-does-mongoose-try-to-create-a-field-with-1-in-element-for-type-map) from January 2019, but with no answers.

**Comment by vkarpov15** [comment]
Your code sample references a `getArchiveId()` function that isn't in your gist, can you clarify what that function does?

Also, the below script executes without error for me. Can you please modify the below script to demonstrate the issue you're seeing?

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7970`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const participantSchema = new Schema({
    userId: {
      required: true,
      type: String,
    },
    role: {
      required: true,
      type: String,
    },
    joinedAt: {
      type: Date,
    },
    rating: {
      type: Number,
    },
  }, {_id: false});

  const sessionSchema = new Schema({
    status: {
      required: true,
      type: String,
      enum: ['requested', 'confirmed', 'started', 'cancelled', 'finished', 'expired'],
    },
    participants: [participantSchema],
    from: {
      type: Date,
    },
    to: {
      type: Date,
    },
    archiveId: {
      type: String,
    },
  });

  const Session = mongoose.model('Session', sessionSchema);

  await Session.create([
    { status: 'requested', participants: [{ userId: 'test', role: 'test' }] }
  ]);

  // Get every session in the db
  const sessions = await Session.find();

  for (const session of sessions) {
    // Compute archiveId for session
    const archiveId = 'test'; //await getArchiveId();

    if (archiveId) {
      try {
        // save archive ID
        session.archiveId = archiveId;
        await session.save();
      } catch (e) {
        /*
        The error i'm getting is 
        `MongoError: Cannot create field '-1' in element {participants:...}`
        */
        console.error(`Could not save archiveId for session ${session._id}\n ${e}`);
      }
    }
  }
}
```

**Comment by vkarpov15** [comment]
Your code sample references a `getArchiveId()` function that isn't in your gist, can you clarify what that function does?

Also, the below script executes without error for me. Can you please modify the below script to demonstrate the issue you're seeing?

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7970`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const participantSchema = new Schema({
    userId: {
      required: true,
      type: String,
    },
    role: {
      required: true,
      type: String,
    },
    joinedAt: {
      type: Date,
    },
    rating: {
      type: Number,
    },
  }, {_id: false});

  const sessionSchema = new Schema({
    status: {
      required: true,
      type: String,
      enum: ['requested', 'confirmed', 'started', 'cancelled', 'finished', 'expired'],
    },
    participants: [participantSchema],
    from: {
      type: Date,
    },
    to: {
      type: Date,
    },
    archiveId: {
      type: String,
    },
  });

  const Session = mongoose.model('Session', sessionSchema);

  await Session.create([
    { status: 'requested', participants: [{ userId: 'test', role: 'test' }] }
  ]);

  // Get every session in the db
  const sessions = await Session.find();

  for (const session of sessions) {
    // Compute archiveId for session
    const archiveId = 'test'; //await getArchiveId();

    if (archiveId) {
      try {
        // save archive ID
        session.archiveId = archiveId;
        await session.save();
      } catch (e) {
        /*
        The error i'm getting is 
        `MongoError: Cannot create field '-1' in element {participants:...}`
        */
        console.error(`Could not save archiveId for session ${session._id}\n ${e}`);
      }
    }
  }
}
```

**Comment by tevin** [comment]
I know this is closed, but in case someone lands here, I think I figured out why this is happening. 

At some point I changed a field in the sub schema `participantSchema` from a `String` to an `Array`.

With the new schema, updates to documents that already had the field populated with a `String` would fail with the above error (even if the update didn't affect the sub-schema). 

I fixed it by writing a script to migrate documents from the old schema to the new one and now updates work as expected.
